src/arbitrary/_std/string.rs

```diff
@@ -0,0 +1,249 @@
+    type Parameters = StringParam;
+    type Strategy = &'static str;
+
+    /// ## Safety
+    ///
+    /// This implementation panics if the input is not a valid regex proptest
+    /// can handle.
+    fn arbitrary_with(args: Self::Parameters) -> Self::Strategy {
+        args.into()
+    }
+}
+
+macro_rules! dst_wrapped {
+    ($($w: ident),*) => {
+        $(arbitrary!($w<str>, MapInto<StrategyFor<String>, Self>, StringParam;
+            a => any_with::<String>(a).prop_map_into()
+        );)*
+    };
+}
+
+dst_wrapped!(Box, Rc, Arc);
+
+lazy_just!(FromUtf16Error, || String::from_utf16(&[0xD800]).unwrap_err());
+
+// This is a void-like type, it needs to be handled by the user of
+// the type by simply never constructing the variant in an enum or for
+// structs by inductively not generating the struct.
+// The same applies to ! and Infallible.
+// generator!(ParseError, || panic!());
+
+arbitrary!(FromUtf8Error, SFnPtrMap<BoxedStrategy<Vec<u8>>, Self>;
+    static_map(not_utf8_bytes(true), |bs| String::from_utf8(bs).unwrap_err())
+);
+
+/// This strategy produces sequences of bytes that are guaranteed to be illegal
+/// wrt. UTF-8 with the goal of producing a suffix of bytes in the end of
+/// an otherwise legal UTF-8 string that causes the string to be illegal.
+/// This is used primarily to generate the `Utf8Error` type and similar.
+pub(crate) fn not_utf8_bytes(allow_null: bool) -> BoxedStrategy<Vec<u8>> {
+    let prefix = collection::vec(any::<char>(), ..::std::u16::MAX as usize);
+    let suffix = gen_el_bytes(allow_null);
+    (prefix, suffix).prop_map(move |(prefix_bytes, el_bytes)| {
+        let iter = prefix_bytes.iter();
+        let string: String = if allow_null {
+            iter.collect()
+        } else {
+            iter.filter(|&&x| x != '\u{0}').collect()
+        };
+        let mut bytes = string.into_bytes();
+        bytes.extend(el_bytes.into_iter());
+        bytes
+    }).boxed()
+}
+
+/// Stands for "error_length" bytes and contains a suffix of bytes that
+/// will cause the whole string to become invalid UTF-8.
+/// See `gen_el_bytes` for more details.
+#[derive(Debug)]
+enum ELBytes {
+    B1([u8; 1]),
+    B2([u8; 2]),
+    B3([u8; 3]),
+    B4([u8; 4])
+}
+
+impl<'a> IntoIterator for &'a ELBytes {
+    type Item = u8;
+    type IntoIter = iter::Map<slice::Iter<'a, u8>, fn(&u8) -> u8>;
+    fn into_iter(self) -> Self::IntoIter {
+        use self::ELBytes::*;
+        (match *self {
+            B1(ref a) => a.iter(),
+            B2(ref a) => a.iter(),
+            B3(ref a) => a.iter(),
+            B4(ref a) => a.iter(),
+        }).map(|x| *x)
+    }
+}
+
+// By analysis of run_utf8_validation defined at:
+// https://doc.rust-lang.org/nightly/src/core/str/mod.rs.html#1429
+// we know that .error_len() \in {None, Some(1), Some(2), Some(3)}.
+// We represent this with the range [0..4) and generate a valid
+// sequence from that.
+fn gen_el_bytes(allow_null: bool) -> BoxedStrategy<ELBytes> {
```

Comment:
Thought: Could this (and its caller `non_utf8_bytes`) be expressed with a [`bytes_regex`](https://docs.rs/proptest/0.4.2/proptest/string/fn.bytes_regex.html)? I.e., something like (not sure about my ranges here)

```
.*(?-ux:
# Two-byte sequence with illegal or missing following byte
[\xC2-\xE0][\x01-\x7F\xC0-\xFF]?|
# More cases...
)
```

Reply:
The language below looks regular to me, so I don't see why not - in fact, haven't I really encoded the regex explicitly? That said, could we revisit this in a later PR perhaps? Changing to a bytes-regex should have no semantic difference, so it wouldn't be a breaking change =)

Reply:
Fair enough

Reply:
I was thinking just now: Should `not_utf8_bytes` and/or `gen_el_bytes` (the latter under a more suitable name...) be exposed as strategies in `proptest::string` or `proptest::char`? I'll make a separate PR for that if so later on =)